File: 0290-word-pattern/0290-word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        
        
		map<string , char> mp1;
		map<char , char> mp2;

		vector<string> v1;
		string temp = "";

		for(int i = 0 ; i < (int)s.length() ; i ++ ){
			if(s[i] == ' ')
			{
				v1.push_back(temp);
				temp = "";
			}else
				temp += s[i];
		}

		if((int)temp.length() != 0)
			v1.push_back(temp);

		int n = (int)v1.size();
		int m = (int)pattern.length();
		if (n != m)
			return false;

		char k = 'a' , k2 = 'a';

		for(int i = 0 ; i < n ; i ++){
			if(mp1.count(v1[i]) > 0 )
				continue;
			else
			{
				mp1[v1[i]] = k;
				k++;
			}

			if(mp2.count(pattern[i]) > 0)
				continue;

			else
			{
				mp2[pattern[i]] = k2;
				k2++;
			}

		}


		for(int i = 0 ; i < n ; i ++ ){
			char x = mp1[v1[i]];
			char y = mp2[pattern[i]];

			if(x != y)
				return false;
		}

	return true;
        
    }
};

File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
            words = s.split(" ")

            if len(words) != len(pattern):
                return False

            d1 = {}
            d2 = {}
            k1, k2 = 'a', 'a'


            for i in range(len(words)):
                if words[i] not in d1:
                    d1[words[i]] = k1
                    k1 = chr(ord(k1) + 1)
                if pattern[i] not in d2:
                    d2[pattern[i]] = k2
                    k2 = chr(ord(k2) + 1)

            for i in range(len(words)):
                if d1[words[i]] != d2[pattern[i]]:
                    return False

            return True
